File: Intersection_LinkedLists.java
// Time Complexity :O(m+n)-where m and n are nodes of the linked lists
// Space Complexity :O(1)
// Did this code successfully run on Leetcode :Yes
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach

public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if(headA==null||headB==null){
            return null;
        }
        
        ListNode current = headA;
        int lengthA = 1;
        while(current.next!=null){
            lengthA++;
            current=current.next;
        }
        
        current = headB;
        int lengthB=1;
        while(current.next!=null){
            lengthB++;
            current=current.next;
        }
        
        while(lengthA>lengthB){
            headA=headA.next;
            lengthA--;
        }
        
        while(lengthB>lengthA){
            headB=headB.next;
            lengthB--;
        }
        
        while(headA!=headB){
            headA = headA.next;
            headB = headB.next;
        }
        
        return headA;
    }
}

File: intersection_linked_lists.py
'''
TC:O(m+n)
SC:O(1)
'''



# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA, headB):
        curA, curB = headA, headB
        countA, countB = 0, 0
        while curA:
            curA = curA.next
            countA+=1
        while curB:
            curB = curB.next
            countB+=1
        
        curA, curB = headA, headB
        
        diff = abs(countA-countB)
        
        if countA > countB:
            for _ in range(diff):
                curA = curA.next
        else:
            for _ in range(diff):
                curB = curB.next
        while curA and curB:
            if curA == curB:
                return curA
            else:
                curA = curA.next
                curB = curB.next
        return None